### PR TITLE
Add revision to trainer push_to_hub

### PR DESCRIPTION
**Pull request by teamclouday**
# What does this PR do?

Add revision to trainer's push_to_hub method. See #33481 

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [ ] Did you read the [contributor guideline](https://github.com/huggingface/transformers/blob/main/CONTRIBUTING.md#create-a-pull-request),
      Pull Request section?
- [ ] Was this discussed/approved via a Github issue or the [forum](https://discuss.huggingface.co/)? Please add a link
      to it if that's the case.
- [ ] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/transformers/tree/main/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/transformers/tree/main/docs#writing-source-documentation).
- [ ] Did you write any new necessary tests?

## Who can review?

Anyone in the community is free to review the PR once the tests have passed. Feel free to tag
members/contributors who may be interested in your PR.

### CODE REVIEW

File: src/transformers/trainer.py [hunk 1]
```diff
@@ -4462,6 +4462,7 @@ def push_to_hub(
         commit_message: Optional[str] = "End of training",
         blocking: bool = True,
         token: Optional[str] = None,
+        revision: str = None,
```
Comment (not-lain) [inline src/transformers/trainer.py:4465]:
```suggestion
        revision: Optional[str] = None,
```

File: src/transformers/trainer.py [hunk 2]
```diff
@@ -4474,6 +4475,8 @@ def push_to_hub(
                 Whether the function should return only when the `git push` has finished.
             token (`str`, *optional*, defaults to `None`):
                 Token with write permission to overwrite Trainer's original args.
+            revision (`str`, *optional*):
+                Branch to push the uploaded files to.
```
Comment (not-lain) [inline src/transformers/trainer.py:4479]:
```suggestion
                The git revision to commit from. Defaults to the head of the "main" branch.
```

**Review by not-lain — COMMENTED** [review]
Thanks for adding this

**Comment by teamclouday** [comment]
Thank you for the suggestions! Just made a new commit

**Comment by not-lain** [comment]
perfect, the failing tests are unrelated to this PR.
cc @SunMarc for review

**Review by SunMarc — APPROVED** [review]
Thanks for adding this @teamclouday  ! Did you test the PR ?

**Comment by teamclouday** [comment]
> Thanks for adding this @teamclouday ! Did you test the PR ?

I just added test code! Can run it with:
```bash
export HUGGINGFACE_CO_STAGING=1; python -m unittest -v tests.trainer.test_trainer.TrainerIntegrationWithHubTester.test_push_to_hub_with_revision
```

Currently with revision, the code is assuming that branch already exists. Wonder if it's necessary to automatically create new branch if not found.

**Comment by not-lain** [comment]
Don't forget to format your files with `ruff`

**Comment by teamclouday** [comment]
> Don't forget to format your files with `ruff`

Thanks for reminder! Just applied the format

**Comment by not-lain** [comment]
the `setup_and_quality` tests are still failing
This can be fixed by organizing the imports and you should be all good.

**Comment by teamclouday** [comment]
Just fixed the imports! Should be good now. Sorry about the back and forth

File: src/transformers/trainer.py [hunk 3]
```diff
@@ -4474,6 +4475,8 @@ def push_to_hub(
                 Whether the function should return only when the `git push` has finished.
             token (`str`, *optional*, defaults to `None`):
                 Token with write permission to overwrite Trainer's original args.
+            revision (`str`, *optional*):
+                The git revision to commit from. Defaults to the head of the "main" branch.
```
Comment (amyeroberts) [inline src/transformers/trainer.py:4479]: Isn't it the branch to commit to - rather than the commit to commit from?

Comment (teamclouday) [reply src/transformers/trainer.py:4479]: I think this description comes from the [hub upload_folder api](https://github.com/huggingface/huggingface_hub/blob/6f49af5051bca567798385b673af76642adf4cf6/src/huggingface_hub/hf_api.py#L4823). I was only going to use it for branches and tested with that, so not sure if it works for other commit hashes.

Comment (amyeroberts) [reply src/transformers/trainer.py:4479]: OK!

**Review by amyeroberts — APPROVED** [review]
Thanks for adding!

**Comment by SunMarc** [comment]
Can you rebase on main to make the CI green @teamclouday ?

**Comment by teamclouday** [comment]
I just rebased to main and CI is green now! Thank you all!

**Pull request merged**